timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/HistoryAnalyzer.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.timelock.corruption.detection;
+
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Iterables;
+import com.palantir.atlasdb.encoding.PtBytes;
+import com.palantir.paxos.PaxosValue;
+import com.palantir.timelock.history.PaxosAcceptorData;
+import com.palantir.timelock.history.models.CompletePaxosHistoryForNamespaceAndUseCase;
+import com.palantir.timelock.history.models.ConsolidatedLearnerAndAcceptorRecord;
+
+public final class HistoryAnalyzer {
+
+    public static boolean runCorruptionCheckOnHistory(CompletePaxosHistoryForNamespaceAndUseCase history) {
+        return verifyLearnersHaveLearnedSameValues(history)
+        && verifyLearnedValueWasAcceptedByQuorum(history)
+        && verifyLearnedValueIsGreatestAcceptedValue(history);
+    }
+
+    @VisibleForTesting
+    static boolean verifyLearnersHaveLearnedSameValues(CompletePaxosHistoryForNamespaceAndUseCase history) {
+        List<ConsolidatedLearnerAndAcceptorRecord> records = history.localAndRemoteLearnerAndAcceptorRecords();
+        return history.getAllSequenceNumbers()
+                .stream()
+                .allMatch(seq -> {
+                    Set<PaxosValue> learnedValuesForRound = getLearnedValuesForRound(records, seq);
+                    return learnedValuesForRound.size() <= 1;
+                });
+    }
+
+    @VisibleForTesting
+    static boolean verifyLearnedValueWasAcceptedByQuorum(CompletePaxosHistoryForNamespaceAndUseCase history) {
+        List<ConsolidatedLearnerAndAcceptorRecord> records = history.localAndRemoteLearnerAndAcceptorRecords();
+        int quorum = getQuorumSize(records);
+
+        return history.getAllSequenceNumbers()
+                .stream()
+                .allMatch(seq -> {
+                    Optional<PaxosValue> optionalLearnedValue = getLearnedValue(records, seq);
+                    if(!optionalLearnedValue.isPresent()) {
+                        return true;
+                    }
+
+                    PaxosValue learnedValue = optionalLearnedValue.get();
+                    List<PaxosValue> acceptedValues = getAcceptedValues(records, seq, learnedValue);
+                    return acceptedValues.size() >= quorum;
```

Comment:
I'd extract a method for this. Depending on taste, you could also process the stream more aggressively and/or use keyed-streams